doc/api/jsonrpc_usage.md

```diff
@@ -0,0 +1,134 @@
+        --negotiate -u : \
+        -X POST \
+        https://$IPAHOSTNAME/ipa/session/login_kerberos
+```
+
+If authentication was successful, `$COOKIEJAR` will contain all session cookies
+returned from the server. We will need to pass this with every request we send
+to the server.
+
+### Password authentication
+
+For password authentication, we just need to post it over HTTPS.
+
+```bash
+$ export COOKIEJAR=/path/to/my.cookie
+$ export IPAHOSTNAME=ipa-master.example.com
+$ s_username=admin s_password=mYSecReT1P2 curl -v  \
+        -H referer:https://$IPAHOSTNAME/ipa  \
+        -H "Content-Type:application/x-www-form-urlencoded" \
+        -H "Accept:text/plain"\
+        -c $COOKIEJAR -b $COOKIEJAR \
+        --cacert /etc/ipa/ca.crt  \
+        --data "user=$s_username&password=$s_password" \
+        -X POST \
+        https://$IPAHOSTNAME/ipa/session/login_password
+```
+
+Same as kerberos authentication, we will need to pass the sessions cookies with
+every request.
+
+### Request and Response format
+
+A JSON-RPC request consists of three properties:
+
+* `method`: A string containing the name of the name that will be called.
+* `params`: the array of parameters for the command.
+* `id`: the request id, it can be of any type, the response will match it.
+
+The response received from the server consists of four properties:
+
+* `result`: The returned Object from the command. If the command failed, this
+  will be null.
+* `principal`: The Kerberos principal under which identity the command was performed.
+* `error`: An Error object containing information about the command if it
+  failed. If it succeeded, this will be null.
+* `ìd`: An ID matching the request this response is replying to.
+
+### Sending a request
+
+Requests should be sent to the API endpoint
+`https://$IPAHOSTNAME/ipa/session/json` over HTTPS. The content type must be set
+to `application/json` and session cookies obtained when authentication must be
+passed with the request.
+
+An example request for the `user_find` command would be:
+
+```bash
+curl -v  \
+	-H referer:https://$IPAHOSTNAME/ipa  \
+        -H "Content-Type:application/json" \
+        -H "Accept:applicaton/json"\
+        -c $COOKIEJAR -b $COOKIEJAR \
+        --cacert /etc/ipa/ca.crt  \
+        -d  '{"method":"user_find","params":[[""],{}],"id":0}' \
+        -X POST \
+        https://$IPAHOSTNAME/ipa/session/json
+```
+
+An easy way to understand how IPA requests are built is via the CLI, by passing
+the `-vv` option to an IPA command.
+
+```bash
+$ ipa -vv user-find 
+ipa: INFO: Request: {
+    "id": 0,
+    "method": "user_find/1",
+    "params": [
+        [],
+        {
+            "version": "2.251"
+        }
+    ]
+}
+
+[...]
+```
+
+## Converting JSON-RPC requests to Python
+
+FreeIPA provides methods to convert JSON-RPC requests to Python format.
+
+```python
+from ipaserver.rpcserver import jsonserver
+
+json_request = '{"method":"user_find","params":[[""],{}],"id":0}'
```

Comment:
This needs `freeipa-python-ipaserver` package (on Fedora). I wonder if we can refactor `ipaserver.rpcserver` and move `marshal()`/`unmarshal()` code into `ipalib.rpc` and call it from there? `unmarshal()` does not depend on `self` at all, `marshal()` only needs a `self.api.env.debug` one that can be passed as a debug argument to that function...

Reply:
Added line about the `freeipa-python-ipaserver` package. I've noted the `rpcserver` refactor for a future PR.